File: RadialMenu/RadialMenuJVM/src/main/java/RadialMenu/RadialMenuItem.java
package RadialMenu;

import java.awt.Color;
import java.util.LinkedList;
import java.util.List;

import com.spvessel.spacevil.ButtonCore;
import com.spvessel.spacevil.CoreWindow;
import com.spvessel.spacevil.ImageItem;
import com.spvessel.spacevil.ItemsLayoutBox;
import com.spvessel.spacevil.Prototype;
import com.spvessel.spacevil.Common.DefaultsService;
import com.spvessel.spacevil.Core.InterfaceBaseItem;
import com.spvessel.spacevil.Core.InterfaceDraggable;
import com.spvessel.spacevil.Core.InterfaceFloating;
import com.spvessel.spacevil.Core.InterfaceFreeLayout;
import com.spvessel.spacevil.Core.InterfaceItem;
import com.spvessel.spacevil.Core.MouseArgs;
import com.spvessel.spacevil.Core.Pointer;
import com.spvessel.spacevil.Decorations.Style;
import com.spvessel.spacevil.Flags.EmbeddedImage;
import com.spvessel.spacevil.Flags.EmbeddedImageSize;
import com.spvessel.spacevil.Flags.KeyCode;
import com.spvessel.spacevil.Flags.LayoutType;

public class RadialMenuItem extends Prototype implements InterfaceFloating, InterfaceFreeLayout, InterfaceDraggable {

    public int radius = 200;
    public int itemRadius = 30;
    public ButtonCore hideButton = new ButtonCore();
    public ButtonCore addButton = new ButtonCore("Add");

    public RadialMenuItem(CoreWindow wnd) {
        ItemsLayoutBox.addItem(wnd, this, LayoutType.FLOATING);

        setPassEvents(false);
        setStyle(DefaultsService.getDefaultStyle(RadialMenuItem.class));
    }

    private boolean _isInit = false;

    @Override
    public void initElements() {
        super.addItem(hideButton);
        super.addItem(addButton);

        ImageItem icon = new ImageItem(
                DefaultsService.getDefaultImage(EmbeddedImage.ARROW_LEFT, EmbeddedImageSize.SIZE_64X64));
        icon.setRotationAngle(90);
        icon.keepAspectRatio(true);
        icon.setColorOverlay(Color.BLACK);

        hideButton.addItem(icon);

        // Events
        hideButton.eventMouseClick.add((sender, args) -> {
            hide();
        });
        eventScrollUp.add((sender, args) -> {
            _degreeDiff = -_scrollStep * _toRadianCoeff;
            rearrangeContacts();
        });
        eventScrollDown.add((sender, args) -> {
            _degreeDiff = _scrollStep * _toRadianCoeff;
            rearrangeContacts();
        });
        eventKeyPress.add((sender, args) -> {
            if (args.key == KeyCode.MENU)
                hide();
        });
        eventMousePress.add(this::onMousePress);
        eventMouseDrag.add(this::onMouseDrag);
        eventMouseDrop.add(this::onMouseDrop);

        for (Contact item : _contacts) {
            super.addItem(item);
            item.updateSize(itemRadius);
        }

        _isInit = true;
    }

    // IDraggable
    private void onMousePress(InterfaceItem sender, MouseArgs args) {
        _beginDegree = calculateDegreeByPos(args.position);
    }

    private void onMouseDrag(InterfaceItem sender, MouseArgs args) {
        double dragDegree = calculateDegreeByPos(args.position);
        _degreeDiff = dragDegree - _beginDegree;
        _beginDegree = dragDegree;
        rearrangeContacts();
    }

    private void onMouseDrop(InterfaceItem sender, MouseArgs args) {
        _degreeDiff = 0;
    }

    // Our own container's AddItem
    @Override
    public void addItem(InterfaceBaseItem item) {
        if (item instanceof Contact) {
            Contact contact = (Contact) item;
            _contacts.add(contact);
            if (_isInit)
                super.addItem(contact);
            contact.updateSize(itemRadius);
        }
        calculateAlphaStep();
        updateLayout();
    }

    @Override
    public boolean removeItem(InterfaceBaseItem item) {
        if (super.removeItem(item)) {
            _contacts.remove(item);
            calculateAlphaStep();
            updateLayout();
            return true;
        }
        return false;
    }

    public List<Contact> getContacts() {
        return new LinkedList<>(_contacts);
    }

    // Style override
    @Override
    public void setStyle(Style style) {
        if (style == null)
            return;
        super.setStyle(style);

        Style innerStyle = style.getInnerStyle("hidebutton");
        hideButton.setStyle(innerStyle);

        innerStyle = style.getInnerStyle("addbutton");
        addButton.setStyle(innerStyle);
    }

    // IFree
    @Override
    public void setHeight(int height) {
        super.setHeight(height);
        updateLayout();
    }

    @Override
    public void setWidth(int width) {
        super.setWidth(width);
        updateLayout();
    }

    private boolean _isUpdating = false;

    // @Override
    public void updateLayout() {
        if (_isUpdating)
            return;
        _isUpdating = true;
        updateControlButtonsPosition();
        rearrangeContacts();
        _isUpdating = false;
    }

    // IFloating
    public void hide() {
        setVisible(false);
        getHandler().setFocus();
    }

    public void show() {
        if (!_isInit) {
            initElements();
            updateLayout();
        }
        setVisible(true);
    }

    public void show(InterfaceItem sender, MouseArgs args) {
        show();
    }

    private boolean _ouside = false;

    public boolean isOutsideClickClosable() {
        return _ouside;
    }

    public void setOutsideClickClosable(boolean value) {
        _ouside = value;
    }

    // logic
    private final double _doubledPI = 2 * Math.PI;
    private final double _toRadianCoeff = Math.PI / 180.0;
    private final double _scrollStep = 10.0;
    private double _alpha = Math.PI;
    private double _beginDegree = 0;
    private double _degreeDiff = 0;
    private List<Contact> _contacts = new LinkedList<Contact>();
    private double _alphaStep = 0;

    private void updateControlButtonsPosition() {
        int x = -hideButton.getWidth() / 2;
        int y = -hideButton.getHeight() / 2;

        x = getWidth() / 2 + getX() + x;
        y = getHeight() / 2 + getY() + y;

        hideButton.setPosition(x, y);
        addButton.setPosition(x + hideButton.getWidth() / 2 - addButton.getWidth() / 2,
                y + hideButton.getHeight() + 10);
    }

    private void rearrangeContacts() {
        Pointer centerPoint = getCenter();
        _alpha += _degreeDiff;

        for (Contact item : _contacts) {
            int x = (int) (radius * Math.cos(_alpha)) + centerPoint.getX() - itemRadius;
            int y = (int) (radius * Math.sin(_alpha)) + centerPoint.getY() - itemRadius;
            item.setPosition(x, y);
            item.setConfines();
            _alpha += _alphaStep;
        }
    }

    private double calculateDegreeByPos(Pointer position) {
        Pointer centerPoint = getCenter();
        double degree = Math.atan2(position.getY() - centerPoint.getY(), position.getX() - centerPoint.getX());
        if (degree < 0)
            degree += _doubledPI;
        return degree;
    }

    private void calculateAlphaStep() {
        _alphaStep = _doubledPI / _contacts.size();
    }

    private Pointer getCenter() {
        Pointer center = new Pointer();
        int xCenter = hideButton.getWidth() / 2 + hideButton.getX();
        int yCenter = hideButton.getHeight() / 2 + hideButton.getY();
        center.setPosition(xCenter, yCenter);
        return center;
    }
}

File: RadialMenu/RadialMenuNet/RadialMenuItem.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using SpaceVIL;
using SpaceVIL.Common;
using SpaceVIL.Core;
using SpaceVIL.Decorations;

namespace RadialMenu
{
    public class RadialMenuItem : Prototype, IFloating, IFreeLayout, IDraggable
    {
        public Int32 Radius = 200;
        public Int32 ItemRadius = 30;
        public ButtonCore HideButton = new ButtonCore();
        public ButtonCore AddButton = new ButtonCore("Add");

        public RadialMenuItem(CoreWindow wnd)
        {
            ItemsLayoutBox.AddItem(wnd, this, LayoutType.Floating);

            SetPassEvents(false);
            SetStyle(DefaultsService.GetDefaultStyle(typeof(RadialMenu.RadialMenuItem)));
        }

        private bool _isInit = false;

        public override void InitElements()
        {
            base.AddItem(HideButton);
            base.AddItem(AddButton);

            ImageItem icon = new ImageItem(
                DefaultsService.GetDefaultImage(EmbeddedImage.ArrowLeft, EmbeddedImageSize.Size64x64));
            icon.SetRotationAngle(90);
            icon.KeepAspectRatio(true);
            icon.SetColorOverlay(Color.Black);

            HideButton.AddItem(icon);

            // Events
            HideButton.EventMouseClick += (sender, args) =>
            {
                Hide();
            };

            EventScrollUp += (sender, args) =>
            {
                _degreeDiff = -_scrollStep * _toRadianCoeff;
                RearrangeContacts();
            };
            EventScrollDown += (sender, args) =>
            {
                _degreeDiff = _scrollStep * _toRadianCoeff;
                RearrangeContacts();
            };
            EventKeyPress += (sender, args) =>
            {
                if (args.Key == KeyCode.Menu)
                    Hide();
            };
            EventMousePress += OnMousePress;
            EventMouseDrag += OnMouseDrag;
            EventMouseDrop += OnMouseDrop;

            foreach (var item in _contacts)
            {
                base.AddItem(item);
                item.UpdateSize(ItemRadius);
            }

            _isInit = true;
        }

        //IDraggable
        private void OnMousePress(IItem sender, MouseArgs args)
        {
            _beginDegree = CalculateDegreeByPos(args.Position);
        }

        private void OnMouseDrag(IItem sender, MouseArgs args)
        {
            double dragDegree = CalculateDegreeByPos(args.Position);
            _degreeDiff = dragDegree - _beginDegree;
            _beginDegree = dragDegree;
            RearrangeContacts();
        }

        private void OnMouseDrop(IItem sender, MouseArgs args)
        {
            _degreeDiff = 0;
        }

        public override void AddItem(IBaseItem item)
        {
            Contact contact = item as Contact;
            if (contact != null)
            {
                _contacts.Add(contact);
                if (_isInit)
                    base.AddItem(contact);
                contact.UpdateSize(ItemRadius);
            }
            CalculateAlphaStep();
            UpdateLayout();
        }

        public override bool RemoveItem(IBaseItem item)
        {
            if (base.RemoveItem(item))
            {
                Contact contact = item as Contact;
                if (contact != null)
                    _contacts.Remove(contact);
                CalculateAlphaStep();
                UpdateLayout();
                return true;
            }
            return false;
        }

        public List<Contact> GetContacts()
        {
            return new List<Contact>(_contacts);
        }

        // Style override
        public override void SetStyle(Style style)
        {
            if (style == null)
                return;
            base.SetStyle(style);

            Style innerStyle = style.GetInnerStyle("hidebutton");
            HideButton.SetStyle(innerStyle);

            innerStyle = style.GetInnerStyle("addbutton");
            AddButton.SetStyle(innerStyle);
        }

        //IFree
        public override void SetHeight(int height)
        {
            base.SetHeight(height);
            UpdateLayout();
        }

        public override void SetWidth(int width)
        {
            base.SetWidth(width);
            UpdateLayout();
        }

        private bool _isUpdating = false;

        public void UpdateLayout()
        {
            if (_isUpdating)
                return;
            _isUpdating = true;
            UpdateControlButtonPosition();
            RearrangeContacts();
            _isUpdating = false;
        }

        //IFloating
        public void Hide()
        {
            SetVisible(false);
            GetHandler().SetFocus();
        }
        
        public void Hide(MouseArgs args)
        {
            Hide();
        }

        public void Show()
        {
            if (!_isInit)
            {
                InitElements();
                UpdateLayout();
            }
            SetVisible(true);
        }

        public void Show(IItem sender, MouseArgs args)
        {
            Show();
        }

        private bool _ouside = false;
        public bool IsOutsideClickClosable()
        {
            return _ouside;
        }
        public void SetOutsideClickClosable(bool value)
        {
            _ouside = value;
        }

        //Model logic
        private readonly double _doubledPI = 2 * Math.PI;
        private readonly double _toRadianCoeff = Math.PI / 180.0;
        private readonly double _scrollStep = 10.0;
        private double _alpha = Math.PI;
        private double _beginDegree = 0;
        private double _degreeDiff = 0;
        private List<Contact> _contacts = new List<Contact>();
        private double _alphaStep = 0;

        private void UpdateControlButtonPosition()
        {
            int x = -HideButton.GetWidth() / 2;
            int y = -HideButton.GetHeight() / 2;

            x = GetWidth() / 2 + GetX() + x;
            y = GetHeight() / 2 + GetY() + y;

            HideButton.SetPosition(x, y);
            AddButton.SetPosition(x + HideButton.GetWidth() / 2 - AddButton.GetWidth() / 2,
                y + HideButton.GetHeight() + 10);
        }

        private void RearrangeContacts()
        {
            Pointer centerPoint = GetCenter();
            _alpha += _degreeDiff;

            foreach (var item in _contacts)
            {
                int x = (int)(Radius * Math.Cos(_alpha)) + centerPoint.GetX() - ItemRadius;
                int y = (int)(Radius * Math.Sin(_alpha)) + centerPoint.GetY() - ItemRadius;
                item.SetPosition(x, y);
                item.SetConfines();
                _alpha += _alphaStep;
            }
        }

        private double CalculateDegreeByPos(Pointer position)
        {
            Pointer centerPoint = GetCenter();
            double degree = Math.Atan2(position.GetY() - centerPoint.GetY(), position.GetX() - centerPoint.GetX());
            if (degree < 0)
                degree += _doubledPI;
            return degree;
        }

        private void CalculateAlphaStep()
        {
            _alphaStep = _doubledPI / _contacts.Count;
        }

        private Pointer GetCenter()
        {
            Pointer center = new Pointer();
            int xCenter = HideButton.GetWidth() / 2 + HideButton.GetX();
            int yCenter = HideButton.GetHeight() / 2 + HideButton.GetY();
            center.SetPosition(xCenter, yCenter);
            return center;
        }
    }
}
